### ISSUE TITLE
Support ` --prefer-binary` in `requirements.txt` files

### ISSUE THREAD

**Issue by stinodego** [astral-sh/uv#1794]
`pip` supports including a `--prefer-binary` flag in a requirements.txt like so:
```
--prefer-binary

polars
ruff
```

However, `uv pip install` does not support this:

```
$ uv pip install -r requirements.txt 
error: Unexpected '-', expected '-c', '-e', '-r' or the start of a requirement in `requirements.txt` at position 0
```

This is on version `0.1.6`.

It would be nice if this syntax could be supported!

**Comment by charliermarsh** [comment]
I think we support `--no-binary` and `--only-binary` on the command-line, but not `--prefer-binary` (and none are supported in requirements.txt right now).

**Comment by stinodego** [comment]
> We already always prefer binaries, right? I believe that's why we omitted the flag.

I'm not sure what `uv` does. I know that `pip` will prefer to install the latest version, even if there is no binary for it. Using `--prefer-binary` will make it install an older version if that means it doesn't have to build from source.

> I'm not sure these belong in the requirements file. Just wondering, why is this preferred (other than compatibility)?

Honestly, I also don't know if it belongs there. I reported this because `uv` wants to be a drop-in replacement for `pip` and I ran into this.

In general though, the benefit of having this setting in the requirements file itself is that you don't have to specify this setting in all your installers, e.g. CI / Makefile / Dockerfile / ... . The setting is in one place, together with the requirements. That does make some sense.

**Comment by charliermarsh** [comment]
@zanieb -- I think `--prefer-binary` goes even further, in that it will prefer an older version of a package with a wheel over a newer version without one.

**Comment by njzjz** [comment]
I am +1 for this feature.  A use case is that h5py linux_aarch64 wheels are available in v3.10 but not available in v3.11 (see https://github.com/h5py/h5py/issues/2408). When I use pip, I can set `export PIP_PREFER_BINARY=1` in the CI service to let pip install v3.10 wheels. However, ux doesn't have such an option. The default behavior is to install from v3.11 source, which will fail as it requires HDF5 library to be installed in advance.

**Comment by zanieb** [comment]
@njzjz Is there a reason that this flag is better than just adding a pin or additional constraint?

**Comment by stinodego** [comment]
You may not want to pin a dependency to take advantage of the latest updates whenever they come out. However, you generally also want to avoid compiling packages from source if binaries are available. This setting unifies the two.

Constraints may not work, or I would have to micromanage these for each dependency / platform depending on the wheels that are available. A single `--prefer-binary` takes care of all that.

**Comment by njzjz** [comment]
This flag will also ensure the CI is not broken in the future when a new release doesn't contain the binary. Otherwise, we need to add the pin manually every time.

**Comment by notatallshaw-gts** [comment]
My use case for `--prefer-binary` is in the equivalent of a `pip compile` scenario, I am generating new pinned versions from my requirements.

Ideally I would prefer `--only-binary` but there are a handful of packages where this is still not possible, and it's not practical to go through my ~200 transitive dependencies and see which ones have not yet releases a wheel for their latest release every time I updated the pinned versions.

**Comment by notatallshaw-gts** [comment]
May be helpful for other users, my use case is now solved by https://github.com/astral-sh/uv/issues/4063 / https://github.com/astral-sh/uv/pull/4067.

I can specify that wheels should always be picked with `--only-binary :all:` and override that to have specific packages use sdists with `--no-binary <package>`

**Comment by njzjz** [comment]
> I can specify that wheels should always be picked with `--only-binary :all:` and override that to have specific packages use sdists with `--no-binary <package>`

I created #4291 for the environment variable support request.

**Comment by tpgillam** [comment]
I agree that this would be useful for compatibility purposes; I'm setting up the matplotlib test environment, and was hoping to use `uv pip sync`. However it doesn't work due to the `--prefer-binary` here:

https://github.com/matplotlib/matplotlib/blob/109ab93ba2e687a7b805cb0075efa52ef4e61cde/requirements/testing/extra.txt#L3

```
> uv pip sync requirements/dev/dev-requirements.txt
error: Error parsing included file in `requirements/dev/dev-requirements.txt` at position 79
  Caused by: Unexpected '-', expected '-c', '-e', '-r' or the start of a requirement at requirements/dev/../testing/extra.txt:3:1
```

**Comment by notatallshaw-gts** [comment]
It would be good if it was in the pip the compatibility document: https://github.com/astral-sh/uv/blob/main/PIP_COMPATIBILITY.md

I guess anyone should submit a PR for that?
